### PR TITLE
Fix indexedFaceSet Ccw flag

### PR DESCRIPTION
**Pull request by BenjaminHug8**
**Description**
The PR #4243 has removed the copy of meshes with cw triangles. Unfortunately the normals are computed before applying `scale` or the `ccw` flag, resulting in wrong computation of colors in shaders.

**Related Issues**
This pull-request fixes issue #4287.

This PR also add the possibility to setup `bool` variables in shaders from wren.

### CODE REVIEW

File: resources/wren/shaders/pbr.frag [hunk 1]
```diff
@@ -209,9 +210,8 @@ vec3 PBRpass(vec3 l, vec3 n, vec3 v, vec3 h, vec4 lightColorAndIntensity, float
 }
 
 void main() {
-  vec3 viewFragmentNormal = normalize(fragmentNormal);
-
-  fragNormal = vec4(normalize(viewFragmentNormal), 1.0) * 0.5 + 0.5;
+  vec3 viewFragmentNormal = (reverseNormals > 0) ? normalize(-fragmentNormal) : normalize(fragmentNormal);
+  fragNormal = vec4(normalize(viewFragmentNormal), 1.0) * 1.0;
```
Comment (omichel) [inline resources/wren/shaders/pbr.frag:214]: This seems odd, why should we multiply by 1.0? It seems useless. Are you sure this change is correct?

File: resources/wren/shaders/pbr_stencil_ambient_emissive.frag [hunk 2]
```diff
@@ -102,8 +103,8 @@ vec3 getIBLContribution(IBLInfo iblInputs, vec3 n, vec3 reflection) {
 }
 
 void main() {
-  vec3 viewFragmentNormal = normalize(fragmentNormal);
-  fragNormal = vec4(normalize(viewFragmentNormal), 1.0) * 0.5 + 0.5;
+  vec3 viewFragmentNormal = (reverseNormals > 0) ? normalize(-fragmentNormal) : normalize(fragmentNormal);
+  fragNormal = vec4(normalize(viewFragmentNormal), 1.0) * 1.0;
```
Comment (omichel) [inline resources/wren/shaders/pbr_stencil_ambient_emissive.frag:107]: Same comment here.

File: resources/wren/shaders/phong_stencil_ambient_emissive.frag [hunk 3]
```diff
@@ -66,7 +67,8 @@ vec4 SRGBtoLINEAR(vec4 srgbIn) {
 }
 
 void main() {
-  fragNormal = vec4(normalize(fragmentNormal), 1.0) * 0.5 + 0.5;
+  vec3 fragmentNormalOriented = (reverseNormals > 0) ? normalize(-fragmentNormal) : normalize(fragmentNormal);
+  fragNormal = vec4(normalize(fragmentNormalOriented), 1.0) * 0.5 + 0.5;
```
Comment (omichel) [inline resources/wren/shaders/phong_stencil_ambient_emissive.frag:71]: Simpler this way:
```suggestion
  fragNormal = vec4(normalize((reverseNormals > 0) ? -fragmentNormal : fragmentNormal), 1.0) * 0.5 + 0.5;
```

File: resources/wren/shaders/pbr.frag [hunk 4]
```diff
@@ -209,9 +210,8 @@ vec3 PBRpass(vec3 l, vec3 n, vec3 v, vec3 h, vec4 lightColorAndIntensity, float
 }
 
 void main() {
-  vec3 viewFragmentNormal = normalize(fragmentNormal);
-
-  fragNormal = vec4(normalize(viewFragmentNormal), 1.0) * 0.5 + 0.5;
+  vec3 viewFragmentNormal = (reverseNormals > 0) ? normalize(-fragmentNormal) : normalize(fragmentNormal);
```
Comment (omichel) [inline resources/wren/shaders/pbr.frag:213]: Simpler (the same applies at other places in the code):
```suggestion
  vec3 viewFragmentNormal = normalize((reverseNormals > 0) ? -fragmentNormal : fragmentNormal);
```

File: src/wren/ShadowVolumeCaster.cpp [hunk 5]
```diff
@@ -110,6 +110,12 @@ namespace wren {
     const Mesh *mesh = mRenderable->mesh();
     const std::vector<Mesh::Edge> &edges = mesh->edges();
 
+    // if negative scale xor cw triangles, reverse normals
+    bool reverseNormal = false;
+    glm::vec3 scale = mRenderable->parent()->scale();
+    if (mRenderable->invertFrontFace() != (scale.x * scale.y * scale.z < 0.0))
+      reverseNormal = true;
```
Comment (omichel) [inline src/wren/ShadowVolumeCaster.cpp:117]: Better:
```suggestion
    glm::vec3 scale = mRenderable->parent()->scale();
    const bool revertNormal = (mRenderable->invertFrontFace() != (scale.x * scale.y * scale.z < 0.0));
```

File: src/wren/Renderable.hpp [hunk 6]
```diff
@@ -98,7 +98,7 @@ namespace wren {
     Renderable();
     virtual ~Renderable();
 
-    void setupAndRender(const ShaderProgram *program);
+    void setupAndRender(ShaderProgram *program);
```
Comment (omichel) [inline src/wren/Renderable.hpp:101]: Why did you remove the const qualifiers from these methods?

**Review by omichel — CHANGES_REQUESTED** [review]

File: src/wren/Renderable.hpp (hunk 6 above)
Comment (BenjaminHug8) [reply src/wren/Renderable.hpp]: I needed to remove them for another way to fix this issue, but I can re add these `const` with the current implementation.

File: resources/wren/shaders/pbr.frag (hunk 1 above)
Comment (BenjaminHug8) [reply resources/wren/shaders/pbr.frag]: This was a simply test that should not have been committed, thank you for spotting this!

File: src/wren/Renderable.cpp [hunk 7]
```diff
@@ -222,9 +224,25 @@ namespace wren {
...
-    if (mInvertFrontFace)
+    glm::vec3 scale = parent()->scale();
+    if (mInvertFrontFace) {
       glstate::setFrontFace((frontFaceMode == GL_CCW) ? GL_CW : GL_CCW);
 
+      if (scale.x * scale.y * scale.z >= 0.0) {
+        GLint location = glGetUniformLocation(program->glName(), "reverseNormals");
+        if (location != -1) {
+          glUniform1i(location, 1);
+        }
```
Comment (omichel) [inline src/wren/Renderable.cpp:235]:
```suggestion
        if (location != -1)
          glUniform1i(location, 1);
```

**Review by omichel — CHANGES_REQUESTED** [review]

File: src/wren/ShadowVolumeCaster.cpp [hunk 8]
```diff
@@ -110,6 +110,10 @@ namespace wren {
     const Mesh *mesh = mRenderable->mesh();
     const std::vector<Mesh::Edge> &edges = mesh->edges();
 
+    // if negative scale xor cw triangles, reverse normals
+    glm::vec3 scale = mRenderable->parent()->scale();
```
Comment (omichel) [inline src/wren/ShadowVolumeCaster.cpp:114]:
```suggestion
    const glm::vec3 &scale = mRenderable->parent()->scale();
```

File: src/wren/Renderable.cpp [hunk 9]
```diff
@@ -222,9 +224,23 @@ namespace wren {
 
     // to render cw and ccw meshes
     const unsigned int frontFaceMode = glstate::getFrontFace();
-    if (mInvertFrontFace)
+    glm::vec3 scale = parent()->scale();
```
Comment (omichel) [inline src/wren/Renderable.cpp:227]:
```suggestion
    const glm::vec3 &scale = parent()->scale();
```

**Review by omichel — CHANGES_REQUESTED** [review]
A couple of constness issues.

**Review by omichel — APPROVED** [review]
Thank you.
